sockeye/training.py

```diff
@@ -121,9 +123,15 @@ def sym_gen(seq_lens):
             source_seq_len, target_seq_len = seq_lens
 
             # source embedding
-            (source_embed,
-             source_embed_length,
-             source_embed_seq_len) = self.embedding_source.encode(source, source_length, source_seq_len)
+            if self.config.num_factors == 0:
+                (source_embed,
+                 source_embed_length,
+                 source_embed_seq_len) = self.embedding_source.encode(source, source_length, source_seq_len)
+            else:
+                (source_embed,
+                 source_embed_length,
+                 source_embed_seq_len) = self.embedding_source.encode(source, source_length, source_seq_len, source_factors)
+
```

Comment:
Another way to do this would be to encode factors as additional dimensions on the tensor; then the logic would go into assembling `source` and choose the right subclass for the embedder, and the calls here would be the same. This would probably be cleaner...

Reply:
I think this would be a good idea. If there are no factors, the data iterator should yield source data of `(batch, time, 1)` or `(batch, time)` [the latter is the current]. If there are source factors present it should yield batches with source data of shape `(batch, time,1+num_factors)`.
Since we elevated the embedding operation as a core model operation (by making it a model component along with encoder & decoder), you can implement the embedding encoder class to split on the third dimension if num_factors > 0.